zavod/zavod/runs.py

```diff
@@ -0,0 +1,94 @@
+import os
+import json
+import string
+import secrets
+from typing import Any, List, Iterator, Optional
+from datetime import datetime, timezone
+
+
+class RunID(object):
+    """A class to represent a run ID, which consists of a timestamp
+    and a random tag."""
+
+    def __init__(self, dt: datetime, tag: str) -> None:
+        self.dt: datetime = dt
+        self.tag: str = tag
+
+    @classmethod
+    def new(cls) -> "RunID":
+        now = datetime.now().astimezone(timezone.utc)
+        now = now.replace(tzinfo=None)
+        now = now.replace(microsecond=0)
+        key = [secrets.choice(string.ascii_uppercase) for _ in range(4)]
+        return cls(now, "".join(key))
+
+    @classmethod
+    def from_string(cls, id: str) -> "RunID":
+        if "-" not in id:
+            raise ValueError(f"Invalid run ID: {id}")
+        ts, tag = id.split("-", 1)
+        dt = datetime.strptime(ts, "%Y%m%d%H%M%S")
+        dt = dt.replace(tzinfo=None)
+        return cls(dt, tag)
+
+    @classmethod
+    def from_env(cls, name: str) -> "RunID":
+        id = os.environ.get(name)
+        if id is None:
+            return cls.new()
+        return cls.from_string(id)
+
+    @property
+    def id(self) -> str:
+        return f"{self.dt.strftime('%Y%m%d%H%M%S')}-{self.tag}"
+
+    def __str__(self) -> str:
+        return self.id
+
+    def __repr__(self) -> str:
+        return f"RunID({self.id})"
+
+    def __eq__(self, other: Any) -> bool:
+        return self.id == str(other)
+
+    def __hash__(self) -> int:
+        return hash(self.id)
+
+
+class RunHistory(object):
+    """A class to represent a history of run IDs."""
+
+    LENGTH = 300
+
+    def __init__(self, items: List[RunID]) -> None:
+        self.items = items
+
+    def append(self, run_id: RunID) -> "RunHistory":
+        """Creates a new history with the given RunID appended."""
+        items = list(self.items)
```

Comment:
The idea is to make a copy instead of actually mutating that particular history. I was playing around in the test cases a bit and that seemed to avoid some ambiguity, especially if we end up caching the history from GCS to the context...

Reply:
Alright, so updating the list of `items` is done with `history.items = history.append(run_id)`? But since it only returns up to `self.LENGTH` that won't work.

Reply:
Ah, I see. You essentially can't have a RunHistory longer than `LENGTH` unless you initialize the RunHistory with a longer list.